lib/FusionInventory/Agent/Task/Inventory/Generic/Ipmi/Fru/Controllers.pm

```diff
@@ -0,0 +1,54 @@
+package FusionInventory::Agent::Task::Inventory::Generic::Ipmi::Fru::Controllers;
+
+use strict;
+use warnings;
+
+use parent 'FusionInventory::Agent::Task::Inventory::Module';
+
+use FusionInventory::Agent::Tools;
+use FusionInventory::Agent::Tools::IpmiFru qw(getIpmiFru parseFru);
+
+my $CONTROLLERS = qr/^(?:
+    BP             |
+    PERC           |
+    NDC            |
+    Ethernet Adptr |
+    SAS Ctlr
+)\d*\s+/x;
```

Comment:
I understand this is used to filter out fru print, but you maybe should also associate a controller `TYPE` value to each match. Indeed your code is missing the controller `TYPE` handling. Anyway, are controllers still discovered by other modules ? If yes, this means we may have duplicated controllers.

Reply:
Is there a list of valid types? I can add, but i don't see how it can help filter out duplicates.

Reply:
About type, there isn't such list but you should see some values in other controllers modules, just search the code for `CONTROLLERS` to find which ones.

Sorry the sentence about duplicate was not related to TYPE. My question should have been, did you check and compare the resulting XML to XML without your patch on some computers ? I think your code may add controllers that has still been seen before in another module. In that case, you may also want to better update with missed datas.

Reply:
A point about TYPE, I'm not sure it is used in plugin and GLPI but it could be in the future. I just want you to check if this can be easy to identify the TYPE. Maybe you'll simply got the type with other modules if you manage to merge this controllers.

Reply:
Duplicates are a big problem when using `lspci` / `getPCIDevices()`. `lspci` may report each chipset on a PCI device individually. Here first four entries are a single onboard NIC and only the last device provides some real data when dug deeper:
```
# lspci | grep Ether
03:00.0 Ethernet controller: Broadcom Corporation NetXtreme II BCM5709 Gigabit Ethernet (rev 20)
03:00.1 Ethernet controller: Broadcom Corporation NetXtreme II BCM5709 Gigabit Ethernet (rev 20)
04:00.0 Ethernet controller: Broadcom Corporation NetXtreme II BCM5709 Gigabit Ethernet (rev 20)
04:00.1 Ethernet controller: Broadcom Corporation NetXtreme II BCM5709 Gigabit Ethernet (rev 20)
06:00.0 Ethernet controller: Mellanox Technologies MT26448 [ConnectX EN 10GigE, PCIe 2.0 5GT/s] (rev b0)

# lspci -s 06:00.0 -vv | grep -a -A 8 'Vital Product Data'
        Capabilities: [48] Vital Product Data
                Product Name: HP ConnectX-2 SFP+
                Read-only fields:
                        [PN] Part number: 666172-001           
                        [EC] Engineering changes: A1
                        [SN] Serial number: IL214601RR              
                        [V0] Vendor specific: PCIe Gen2 x8    
                        [RV] Reserved: checksum good, 0 byte(s) reserved
                Read/write fields:

# lspci -s 03:00.0 -vv | grep -A 8 'Vital Product Data'
        Capabilities: [50] Vital Product Data
                Product Name: HP NC382i Multifunction Gigabit Server Adapter
                Read-only fields:
                        [PN] Part number: N/A
                        [EC] Engineering changes: N/A
                        [SN] Serial number: 0123456789
                        [MN] Manufacture ID: 31 30 33 43
                        [RV] Reserved: checksum good, 37 byte(s) reserved
                End

```

Or another host `lspci` provides somewhat sensible data, but in reality all four controllers below are a single NIC with a completely different part number:

```
# lspci | grep Ether
01:00.0 Ethernet controller: Broadcom Limited NetXtreme BCM5720 Gigabit Ethernet PCIe
01:00.1 Ethernet controller: Broadcom Limited NetXtreme BCM5720 Gigabit Ethernet PCIe
02:00.0 Ethernet controller: Broadcom Limited NetXtreme BCM5720 Gigabit Ethernet PCIe
02:00.1 Ethernet controller: Broadcom Limited NetXtreme BCM5720 Gigabit Ethernet PCIe

# lspci -s 01:00.0 -vv | grep -a -A 8 'Vital Product Data'
        Capabilities: [50] Vital Product Data
                Product Name: Broadcom NetXtreme Gigabit Ethernet
                Read-only fields:
                        [PN] Part number: BCM95720
                        [MN] Manufacture ID: 31 30 32 38
                        [V0] Vendor specific: FFV20.2.17
                        [V1] Vendor specific: DSV1028VPDR.VER1.0
                        [V2] Vendor specific: NPY2
                        [V3] Vendor specific: PMT1

# ipmitool fru list | grep -A 5 ': NDC'
FRU Device Description : NDC (ID 4)
 Board Mfg Date        : Sat Oct  5 02:28:00 2013
 Board Mfg             : DELL
 Board Product         : BRCM GbE 4P 5720-t rNDC       
 Board Serial          : CN754353A2025Q
 Board Part Number     : 0FM487A02
```

I would like to leave `lspci` as a last resort source of controllers, but it's needed to get some NICs, GPUs (except for internal ones (?)) and probably other devices.

Reply:
I understand you're complaining many lspci entries are associated with only one physical device. Btw remember we are listing controllers, I mean hardware system entries to access physical devices. Maybe we are starting to discuss about a confusing subject and the inventory may be not clear about that but I don't think this is the right place to open this discussion.

Anyway my concern about duplicate: it seems to me you're adding controllers that may still have been inventoried in other modules. My question is then do you think you can deduplicate/update the still found controllers ?

Reply:
I'm afraid i can't. Other modules don't provide a unique field (serial) i could filter on.